File: 0962 maximum-width-ramp/solution.go
func maxWidthRamp(nums []int) int {
    n := len(nums)
    stack := make([]int, 0)
    for i := 0; i < n; i++ {
        if len(stack) == 0 || nums[i] < nums[stack[len(stack)-1]] {
            stack = append(stack, i)
        }
    }
    ans := 0
    for j := n - 1; j >= 0; j-- {
        for len(stack) > 0 && nums[j] >= nums[stack[len(stack)-1]] {
            i := stack[len(stack)-1]
            if j-i > ans {
                ans = j - i
            }
            stack = stack[:len(stack)-1]
        }
    }
    return ans
}

File: 0962 maximum-width-ramp/solution.rs
impl Solution {
    pub fn max_width_ramp(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut stack: Vec<usize> = Vec::new();
        for i in 0..n {
            if stack.is_empty() || nums[i] < nums[*stack.last().unwrap()] {
                stack.push(i);
            }
        }
        let mut ans = 0;
        for j in (0..n).rev() {
            while let Some(&i) = stack.last() {
                if nums[j] >= nums[i] {
                    ans = ans.max(j - i);
                    stack.pop();
                } else {
                    break;
                }
            }
        }
        ans as i32
    }
}

